zcash_primitives/src/zip32/sapling.rs

```diff
@@ -0,0 +1,1811 @@
+    }
+
+    /// Exposes the external [`FullViewingKey`] component of this diversifiable full viewing key.
+    pub fn fvk(&self) -> &FullViewingKey {
+        &self.fvk
+    }
+
+    /// Derives a nullifier-deriving key for the provided scope.
+    ///
+    /// This API is provided so that nullifiers for change notes can be correctly computed.
+    pub fn to_nk(&self, scope: Scope) -> NullifierDerivingKey {
+        match scope {
+            Scope::External => self.fvk.vk.nk,
+            Scope::Internal => self.derive_internal().fvk.vk.nk,
+        }
+    }
+
+    /// Derives an incoming viewing key corresponding to this full viewing key.
+    pub fn to_ivk(&self, scope: Scope) -> SaplingIvk {
+        match scope {
+            Scope::External => self.fvk.vk.ivk(),
+            Scope::Internal => self.derive_internal().fvk.vk.ivk(),
+        }
+    }
+
+    /// Derives an outgoing viewing key corresponding to this full viewing key.
+    pub fn to_ovk(&self, scope: Scope) -> OutgoingViewingKey {
+        match scope {
+            Scope::External => self.fvk.ovk,
+            Scope::Internal => self.derive_internal().fvk.ovk,
+        }
+    }
+
+    /// Attempts to produce a valid payment address for the given diversifier index.
+    ///
+    /// Returns `None` if the diversifier index does not produce a valid diversifier for
+    /// this `DiversifiableFullViewingKey`.
+    pub fn address(&self, j: DiversifierIndex) -> Option<PaymentAddress> {
+        sapling_address(&self.fvk, &self.dk, j)
+    }
+
+    /// Finds the next valid payment address starting from the given diversifier index.
+    ///
+    /// This searches the diversifier space starting at `j` and incrementing, to find an
+    /// index which will produce a valid diversifier (a 50% probability for each index).
+    ///
+    /// Returns the index at which the valid diversifier was found along with the payment
+    /// address constructed using that diversifier, or `None` if the maximum index was
+    /// reached and no valid diversifier was found.
+    pub fn find_address(&self, j: DiversifierIndex) -> Option<(DiversifierIndex, PaymentAddress)> {
+        sapling_find_address(&self.fvk, &self.dk, j)
+    }
+
+    /// Returns the payment address corresponding to the smallest valid diversifier index,
+    /// along with that index.
+    pub fn default_address(&self) -> (DiversifierIndex, PaymentAddress) {
+        sapling_default_address(&self.fvk, &self.dk)
+    }
+
+    /// Returns the internal address corresponding to the smallest valid diversifier index,
+    /// along with that index.
+    ///
+    /// This address **MUST NOT** be encoded and exposed to end users. User interfaces
+    /// should instead mark these notes as "change notes" or "internal wallet operations".
+    pub fn change_address(&self) -> (DiversifierIndex, PaymentAddress) {
+        let internal_dfvk = self.derive_internal();
+        sapling_default_address(&internal_dfvk.fvk, &internal_dfvk.dk)
+    }
+
```

Comment:
Should there be a `pub fn find_change_address(&self, j: DiversifierIndex) -> Option<(DiversifierIndex, PaymentAddress)`?

Reply:
No; we always use the same internal address; since change is always shielded, and the address is never shared, there's no benefit to using a different diversified address.